### DISCUSSION TITLE
RavenDb 5.4 + C# Records + Predefined Ids = Attempted to associate a different object with id

### DISCUSSION THREAD

**Discussion by JohnGalt1717** [ravendb/ravendb#15291] [Q&A] [answered]
If I do something like this:

var person = new Person {
   Id = Guild.NewGuid().ToString(),
   Name = "Hi!"
};

dbSession.Store(person);
dbSession.SaveChanges();

person = person with {
    Name = "Low!"
};
dbSession.Store(person);
dbSession.SaveChanges();

On the second store I get the above error.

If I instead change the property to set; instead of init; and just set the value, then everything works fine.

Is with {} not supported in RavenDb?

I would have thought that it would have looked it up in whatever is caching before the SaveChangesAsync() by the Id value and replaced it with the new record because the Id values were the same?

**Comment by mcrio** [comment]
I guess the session doesn't like having two objects with the same ID. Try evicting the previous object from the session before storing the new one.

**Reply by JohnGalt1717** [reply]
Why wouldn't it just replace because the Id value is the same on the store?

**Reply by DejanMilicic** [reply]
internally, entity is placed in a dictionary as `<entity.id, entity>`
`session.store(entity2)` will produce error since dictionary already contains entry with existing id

**Comment by DejanMilicic** [comment, marked as answer]
`session.Store` will populate the identity map of a session with a dictionary entry `<entity.id, entity>`
Hence, `session.Store(person)` will create new entry `<person.id, person>`

Records are immutable, so once you execute `person = person with {...}` you created new record.
You can effectively consider this to be `person2 = person with {...}`

So, when you attempt `session.Store(person2)` you are getting exception from store because person2 has same id
but different content as already existing identity map entry.

You can mitigate this by manually evicting record from the identity map
https://github.com/DejanMilicic/RavenWorkshop/blob/main/Northwind/Northwind/Features/Records/Records.cs#L55-L65

or you can load record from the database without tracking - this will not create entry in the identity map
https://github.com/DejanMilicic/RavenWorkshop/blob/main/Northwind/Northwind/Features/Records/Records.cs#L67-L78

**Reply by ayende** [reply]
Consider this scenario:

```
       record Person(string Name) { public int Age {get;set;} };

        var p = new Person("Oren");

       session.Store(p, "person/1");

        var np = p with {Name = "Eini"};
       session.Store(np, "person/1");
        p.Age = 2;
        np.Age = 3;

       session.SaveChanges();
```

What would you *expect* the value of the `Age` property for `person/1` to be?
We have _two_ distinct objects that has the same name.

RavenDB also does unit of work, to ensure that a single instance is mapped ot all references to the same document inside the session.

**Comment by JohnGalt1717** [comment]
I would expect the age to be  3.

Because p, by the rules of immutable types is now a legacy value that is not linked to anything anywhere because .Store replaced person/1 with np and thus unlinked p from the tracking.

That's the way immutable types are supposed to work.

**Reply by ayende** [reply]
Not how the `UnitOfWork` pattern works.
We track _references_ to objects, not their clones.

**Reply by JohnGalt1717** [reply]
But literally .Store replaces the reference with the new one because the Id is the same. The old one is outright tossed because the Id values match. Everything else about it doesn't matter. Just the Id.

In the case of a record you literally can't track references to objects because to mutate the object you have to replace the object. That's the entire point.

**Reply by ayende** [reply]
It is *far* more common to get to a place where you'll have confusion around that.

Consider the following code:

```
record Person(string Name, /* ... others */);

var p = session.Load<Person>(id);
p = p with { Name = "Other" };
session.Store(p);

var people = session.Query<Person>().ToList(); // this returns the document with `id` 

// what do you have now?
```

Note that in this case, you load, modify and store the document, but also query it, loading the _original_ values from the database.
For that reason, we keep a *single* instance per session for each document.

**Reply by JohnGalt1717** [reply]
According to the rules of immutability, your Store() was overwritten by your Query, p is detached but maintained without mutation (i.e. it wasn't reverted back to what was in the database)

Again, this is the entire point to immutable records. At no point can anything change without you meaning it and explicitly creating a new object to do it.

**Reply by ayende** [reply]
You cannot mix immutable and unit of work
You can try making it work for you by setting `DoNotTrack` for the session

**Reply by JohnGalt1717** [reply]
Ok, well I'd suggest then that the documentation be made clear that immutable records shouldn't be used for anything but projections in RavenDb unless you want a ton of work and outline why.  (And I don't know why you'd use non-immutable records when the equality is based on Id so you'd just be increasing overhead)

**Reply by ayende** [reply]
See here for the reasoning: https://martinfowler.com/eaaCatalog/identityMap.html

See here for more discussion on a similar scenario:
https://jeremydmiller.com/2016/03/29/the-identity-map-pattern-in-marten/
